File: app/src/main/java/br/edu/ifspsaocarlos/agenda/contentprovider/ContactProvider.java
package br.edu.ifspsaocarlos.agenda.contentprovider;

import android.content.ContentProvider;
import android.content.ContentUris;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.support.annotation.Nullable;

import br.edu.ifspsaocarlos.agenda.data.SQLiteHelper;

/**
 * Created by rafae on 14/12/2015.
 */
public class ContactProvider extends ContentProvider {
    public static final  int        CONTATOS    = 1;
    public static final  int        CONTATOS_ID = 2;
    private static final UriMatcher sURIMatcher = new UriMatcher(UriMatcher.NO_MATCH);

    static {
        sURIMatcher.addURI(Contatos.AUTHORITY, "contatos", CONTATOS);
        sURIMatcher.addURI(Contatos.AUTHORITY, "contatos/#", CONTATOS_ID);
    }

    private SQLiteDatabase database;
    private SQLiteHelper   dbHelper;

    @Override
    public boolean onCreate() {
        dbHelper = new SQLiteHelper(getContext());
        return true;
    }

    @Nullable
    @Override
    public Cursor query(Uri uri, String[] projection, String selection, String[] selectionArgs, String sortOrder) {

        database = dbHelper.getWritableDatabase();

        Cursor cursor;

        switch (sURIMatcher.match(uri)) {
            case CONTATOS:
                cursor = database.query(SQLiteHelper.DATABASE_TABLE, projection, selection, selectionArgs, null, null, sortOrder);
                break;

            case CONTATOS_ID:
                cursor = database.query(SQLiteHelper.DATABASE_TABLE, projection, Contatos.KEY_ID + "=" + uri.getLastPathSegment(), null,
                        null, null, sortOrder);
                break;

            default:
                throw new IllegalArgumentException("URI desconhecida");
        }
        return cursor;
    }

    @Nullable
    @Override
    public String getType(Uri uri) {

        switch (sURIMatcher.match(uri)) {
            case CONTATOS:
                return Contatos.CONTENT_TYPE;

            case CONTATOS_ID:
                return Contatos.CONTENT_ITEM_TYPE;

            default:
                throw new IllegalArgumentException("URI desconhecida");
        }
    }

    @Nullable
    @Override
    public Uri insert(Uri uri, ContentValues values) {

        database = dbHelper.getWritableDatabase();

        int uriType = sURIMatcher.match(uri);
        long id;

        switch (uriType) {
            case CONTATOS:
                id = database.insert(SQLiteHelper.DATABASE_TABLE, null, values);
                break;

            default:
                throw new IllegalArgumentException("URI desconhecida");

        }

        uri = ContentUris.withAppendedId(uri, id);
        return uri;
    }

    @Override
    public int delete(Uri uri, String selection, String[] selectionArgs) {

        database = dbHelper.getWritableDatabase();

        int uriType = sURIMatcher.match(uri);
        int count;

        switch (uriType) {
            case CONTATOS:
                count = database.delete(SQLiteHelper.DATABASE_TABLE, selection, selectionArgs);
                break;

            case CONTATOS_ID:
                count = database.delete(SQLiteHelper.DATABASE_TABLE, Contatos.KEY_ID + "=" + uri.getPathSegments().get(1), null);
                break;

            default:
                throw new IllegalArgumentException("URI desconhecida");

        }
        database.close();
        return count;
    }

    @Override
    public int update(Uri uri, ContentValues values, String selection, String[] selectionArgs) {
        database = dbHelper.getWritableDatabase();

        int uriType = sURIMatcher.match(uri);
        int count;

        switch (uriType) {
            case CONTATOS:
                count = database.update(SQLiteHelper.DATABASE_TABLE, values, selection, selectionArgs);
                break;

            case CONTATOS_ID:
                count = database.update(SQLiteHelper.DATABASE_TABLE, values, Contatos.KEY_ID + "=" + uri.getPathSegments().get(1), null);
                break;

            default:
                throw new IllegalArgumentException("URI desconhecida");
        }
        database.close();
        return count;
    }

    public static final class Contatos {

        public static final String AUTHORITY = "br.edu.ifspsaocarlos.agenda.provider";


        public static final Uri CONTENT_URI = Uri.parse("content://" + AUTHORITY + "/contatos");

        public static final String CONTENT_TYPE      = "vnd.android.cursor.dir/vnd.br.edu.ifspsaocarlos.agenda.contatos";
        public static final String CONTENT_ITEM_TYPE = "vnd.android.cursor.item/vnd.br.edu.ifspsaocarlos.agenda.contatos";

        public static final String KEY_ID       = "id";
        public static final String KEY_NOME     = "nome";
        public static final String KEY_TELEFONE = "fone";
    }
}

File: app/src/main/java/br/edu/ifspsaocarlos/agenda/contentprovider/ContactProvider.kt
package br.edu.ifspsaocarlos.agenda.contentprovider

import android.content.ContentProvider
import android.content.ContentUris
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.net.Uri
import br.edu.ifspsaocarlos.agenda.data.Database.ContactsTable
import br.edu.ifspsaocarlos.agenda.data.Database.ContactsTable.KEY_ID
import dagger.hilt.android.EntryPointAccessors

class ContactProvider : ContentProvider() {

    private lateinit var database: SQLiteDatabase

    override fun onCreate(): Boolean {
        context?.applicationContext?.let {
            val entryPoint = EntryPointAccessors.fromApplication(it, ContactProviderEntryPoint::class.java)
            database = entryPoint.database
        }

        return true
    }

    override fun query(
        uri: Uri,
        projection: Array<String>?,
        selection: String?,
        selectionArgs: Array<String>?,
        sortOrder: String?
    ): Cursor? {
        val cursor = when (sURIMatcher.match(uri)) {
            CONTACTS -> database.query(
                ContactsTable.TABLE_NAME,
                projection,
                selection,
                selectionArgs,
                null,
                null,
                sortOrder
            )

            CONTACTS_ID -> database.query(
                ContactsTable.TABLE_NAME,
                projection,
                "$KEY_ID = ${uri.lastPathSegment}",
                null,
                null,
                null,
                sortOrder
            )

            else -> throw IllegalArgumentException("Unknown URI")
        }
        return cursor
    }

    override fun getType(uri: Uri): String {
        return when (sURIMatcher.match(uri)) {
            CONTACTS -> Contacts.CONTENT_TYPE

            CONTACTS_ID -> Contacts.CONTENT_ITEM_TYPE

            else -> throw IllegalArgumentException("Unknown URI")
        }
    }

    override fun insert(
        uri: Uri,
        values: ContentValues?
    ): Uri {
        val uriType = sURIMatcher.match(uri)
        val id: Long

        when (uriType) {
            CONTACTS -> id = database.insert(
                ContactsTable.TABLE_NAME,
                null,
                values
            )

            else -> throw IllegalArgumentException("Unknown URI")

        }
        return ContentUris.withAppendedId(uri, id)
    }

    override fun delete(
        uri: Uri,
        selection: String?,
        selectionArgs: Array<String>?
    ): Int {
        val uriType = sURIMatcher.match(uri)

        val count = when (uriType) {
            CONTACTS -> database.delete(
                ContactsTable.TABLE_NAME,
                selection,
                selectionArgs
            )

            CONTACTS_ID -> database.delete(
                ContactsTable.TABLE_NAME,
                "$KEY_ID = ${uri.pathSegments[1]}",
                null
            )

            else -> throw IllegalArgumentException("Unknown URI")

        }
        return count
    }

    override fun update(
        uri: Uri,
        values: ContentValues?,
        selection: String?,
        selectionArgs: Array<String>?
    ): Int {
        val uriType = sURIMatcher.match(uri)

        val count = when (uriType) {
            CONTACTS -> database.update(
                ContactsTable.TABLE_NAME,
                values,
                selection,
                selectionArgs
            )

            CONTACTS_ID -> database.update(
                ContactsTable.TABLE_NAME,
                values,
                "$KEY_ID = ${uri.pathSegments[1]}",
                null
            )

            else -> throw IllegalArgumentException("Unknown URI")
        }
        return count
    }

    object Contacts {

        const val AUTHORITY: String = "br.edu.ifspsaocarlos.agenda.provider"

        val CONTENT_URI: Uri = Uri.parse("content://$AUTHORITY/contacts")

        const val CONTENT_TYPE: String = "vnd.android.cursor.dir/vnd.br.edu.ifspsaocarlos.agenda.contacts"
        const val CONTENT_ITEM_TYPE: String = "vnd.android.cursor.item/vnd.br.edu.ifspsaocarlos.agenda.contacts"
    }

    companion object {

        const val CONTACTS: Int = 1
        const val CONTACTS_ID: Int = 2

        private val sURIMatcher = UriMatcher(UriMatcher.NO_MATCH)

        init {
            with(sURIMatcher) {
                addURI(Contacts.AUTHORITY, "contacts", CONTACTS)
                addURI(Contacts.AUTHORITY, "contacts/#", CONTACTS_ID)
            }
        }
    }
}